### PR TITLE
Creating a new IlliXR Plugin that generates fake pose data

### PR DESCRIPTION
**Pull request by wsherman64**
I created a short new plugin that just generates a fake pose value for the headset tracking using a very simple mathematical algorithm -- the head moves in a circle, always looking forward.

### CODE REVIEW

File: FauxPose/plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	virtual void set_offset(const Eigen::Quaternionf& raw_o_times_offset) override{
+		std::unique_lock lock {offset_mutex};
+		Eigen::Quaternionf raw_o = raw_o_times_offset * offset.inverse();
+		//std::cout << "pose_prediction: set_offset" << std::endl;
+		offset = raw_o.inverse();
+	}
+
+	// ********************************************************************
+	virtual fast_pose_type get_fast_pose() const override  {                                          
+	//	const switchboard::ptr<const switchboard::event_wrapper<time_type>> estimated_vsync = _m_vsync_estimate.get_ro_nullable();
```
Comment (charmoniumQ) [inline FauxPose/plugin.cpp:92]: Delete if not necessary.

File: FauxPose/plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	// NOTE: time_type == std::chrono::system_clock::time_point
+	virtual fast_pose_type get_fast_pose(time_type time) const override {
+		pose_type simulated_pose;
+		double	sim_time;
+
+		RAC_ERRNO_MSG("[fauxpose] at start of _p_one_iteration");
+
+		// Calculate simulation time from start of execution
+		std::chrono::nanoseconds elapsed_time;
+		//auto now = std::chrono::system_clock::now();
```
Comment (charmoniumQ) [inline FauxPose/plugin.cpp:115]: Delete if not necessary.

File: FauxPose/plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+		//printf("[fauxpose] time = %ld\n", std::chrono::high_resolution_clock::now().time_since_epoch().count());
+		printf("[fauxpose] elapsed time = %ld (%lf)\n", elapsed_time.count(), sim_time);
+		//fprintf(stderr, "[fauxpose] ********* elapsed time = %ld (%lf)\n", elapsed_time.count(), sim_time);
+
+		// Calculate new pose values
+		//simulated_location = Eigen::Vector3f{0.1f, 0.0f, 0.0f};
+
+		//fprintf(stderr, "FP: time.count is %ld\n", time.time_since_epoch());
+		simulated_pose.position[0] = amplitude * sin(sim_time * period);	// X
```
Comment (charmoniumQ) [inline FauxPose/plugin.cpp:129]: Note in a comment that `sim_start_time` is just a phase offset in the sine wave.

File: FauxPose/plugin.cpp (hunk 1 above)
Comment (wsherman64) [reply FauxPose/plugin.cpp]: Removed commented out code.

File: FauxPose/plugin.cpp (hunk 2 above)
Comment (wsherman64) [reply FauxPose/plugin.cpp]: Removed commented-out code.

File: FauxPose/plugin.cpp (hunk 3 above)
Comment (wsherman64) [reply FauxPose/plugin.cpp:120]: Added variable descriptions here and elsewhere.

File: FauxPose/plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	// get_fast_pose(): returns a "fast_pose_type" with the algorithmically
+	//   determined location values.  (Presently moving in a circle, but
+	//   always facing "front".)
+	//
+	// NOTE: time_type == std::chrono::system_clock::time_point
+	virtual fast_pose_type get_fast_pose(time_type time) const override {
+		pose_type simulated_pose;	/* The algorithmically calculated 6-DOF pose */
+		double	sim_time;		/* sim_time is used to regulate a consistent movement */
+
+		RAC_ERRNO_MSG("[fauxpose] at start of _p_one_iteration");
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:106]: Remove.

File: FauxPose/plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+		// Calculate simulation time from start of execution
+		std::chrono::nanoseconds elapsed_time;
+		elapsed_time = time - sim_start_time;
+		sim_time = elapsed_time.count() * 0.000000001;
+
+		//printf("[fauxpose] time = %ld\n", std::chrono::high_resolution_clock::now().time_since_epoch().count());
+		printf("[fauxpose] elapsed time = %ld (%lf)\n", elapsed_time.count(), sim_time);
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:114]: Guard with `#ifndef NDEBUG`.

File: FauxPose/plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		// Calculate new pose values
+		//   Pose values are calculated from the passage of time to maintain consistency */
+		//fprintf(stderr, "FP: time.count is %ld\n", time.time_since_epoch());
+		simulated_pose.position[0] = amplitude * sin(sim_time * period);	// X
+		simulated_pose.position[1] = 1.5;					// Y
+		simulated_pose.position[2] = amplitude * cos(sim_time * period);	// Z
+		simulated_pose.orientation = Eigen::Quaternionf(1.0, 0.0, 0.0, 0.0);	// W,X,Y,Z (No rotation)
+
+		// Return the new pose
+		printf("[fauxpose] Returning pose\n");
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:126]: Guard with `#ifndef NDEBUG`.

File: FauxPose/plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+		double	sim_time;		/* sim_time is used to regulate a consistent movement */
+
+		RAC_ERRNO_MSG("[fauxpose] at start of _p_one_iteration");
+
+		// Calculate simulation time from start of execution
+		std::chrono::nanoseconds elapsed_time;
+		elapsed_time = time - sim_start_time;
+		sim_time = elapsed_time.count() * 0.000000001;
+
+		//printf("[fauxpose] time = %ld\n", std::chrono::high_resolution_clock::now().time_since_epoch().count());
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:113]: Remove.

File: FauxPose/plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+		printf("[fauxpose] elapsed time = %ld (%lf)\n", elapsed_time.count(), sim_time);
+		//fprintf(stderr, "[fauxpose] ********* elapsed time = %ld (%lf)\n", elapsed_time.count(), sim_time);
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:115]: Remove.

File: FauxPose/plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+		// Calculate new pose values
+		//   Pose values are calculated from the passage of time to maintain consistency */
+		//fprintf(stderr, "FP: time.count is %ld\n", time.time_since_epoch());
```
Comment (mhuzai) [inline FauxPose/plugin.cpp:119]: Remove.

**Review by mhuzai — COMMENTED** [review]
Looks great overall. Most of the comments just pertain to style and code cleanup.

**Review by jianxiapyh — COMMENTED** [review]
agree with huzi's comment

File: FauxPose/plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+public:
+    // ********************************************************************
+    /* Constructor: Provide handles to faux_pose */
+    faux_pose(const std::string& name, phonebook* pb)
+        : plugin{name, pb} {
+        // "pose_prediction" is a class inheriting from "phonebook::service"
+        //   It is described in "pose_prediction.hpp"
+        pb->register_impl<pose_prediction>(std::static_pointer_cast<pose_prediction>(std::make_shared<faux_pose_impl>(pb)));
+#ifndef NDEBUG
+        printf("[fauxpose] Starting Plugin\n");
```
Comment (qinjunj) [inline FauxPose/plugin.cpp:157]: Better use `std::cout` for consistency.

File: FauxPose/plugin.cpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    // ---------------------------------------------------------------------
+    // ---------------------------------------------------------------------
+    // ---------------------------------------------------------------------
+private:
+    const std::shared_ptr<switchboard> sb;
+    mutable Eigen::Quaternionf         offset{Eigen::Quaternionf::Identity()};
+    mutable std::shared_mutex          offset_mutex;
+
+    time_type sim_start_time; /* Store the initial time to calculate a known runtime */
```
Comment (qinjunj) [inline FauxPose/plugin.cpp:140]: `time_type` is obsolete. Please use `time_point` in `common/relative_clock.hpp`.

Comment (qinjunj) [inline FauxPose/plugin.cpp:140]: `time_type` is obsolete. Please use `time_point` in `common/relative_clock.hpp`.

**Review by jianxiapyh — CHANGES_REQUESTED** [review]
I guess it just need the change of the time_type?

**Comment by Jebbly** [comment]
Closed because an updated PR was merged as of #389

**Pull request closed without merging**